### PR TITLE
fix(proxy): honour `allow_requests_on_db_unavailable` in readiness probe

### LINKED ISSUE #34934: [Bug]: `/health/readiness` returns 503 during a DB outage even with `allow_requests_on_db_unavailable: true`, pulling every pod out of rotation
### Check for existing issues

- [x] I have searched the existing issues and checked that my issue is not a duplicate.

### What happened?

`general_settings.allow_requests_on_db_unavailable: true` is the HA switch that keeps the proxy serving when the database is unreachable, and it works as advertised at the request layer. `PrismaDBExceptionHandler.handle_db_exception` 

https://github.com/BerriAI/litellm/blob/61c7e706dd24db17131053445c36bd00ad8fa228/litellm/proxy/db/exception_handler.py#L262-L262

and 

https://github.com/BerriAI/litellm/blob/48b5a5a0cc5a694a11219416ee0b6eb6e620e74e/litellm/proxy/auth/auth_exception_handler.py#L67-L67

both consult it and fail open

`/health/readiness` never consults it. When a Prisma DB is configured but unreachable, both `_resolve_public_readiness_db` 

https://github.com/BerriAI/litellm/blob/5f864c83ce7b7586e2524f610dd7dc1712f184ac/litellm/proxy/health_endpoints/_health_endpoints.py#L1534-L1548

and 

https://github.com/BerriAI/litellm/blob/5f864c83ce7b7586e2524f610dd7dc1712f184ac/litellm/proxy/health_endpoints/_health_endpoints.py#L1457-L1458

set HTTP 503 unconditionally. There is no `general_settings` key that turns this off

With the shipped Helm defaults,

https://github.com/BerriAI/litellm/blob/a643dd08206183b480f9eba1a6c94de29e045899/helm/litellm-helm/values.yaml#L100-L106

a database outage lasting roughly 30 seconds marks every replica NotReady at once. Kubernetes then pulls all of them out of the Service endpoints, so requests never reach the proxy and the request-layer fail-open never gets a chance to run. The flag that exists specifically to survive a DB outage is defeated by the probe before it can do anything

Two things widen the blast radius. `/health/liveliness` does not look at the DB, so the pods sit NotReady indefinitely without restarting until the database comes back. And `startupProbe.path` also defaults to `/health/readiness` with `failureThreshold: 30`, so any rollout that overlaps a DB outage ends in CrashLoopBackOff rather than a degraded but serving pod

<details><summary>Details</summary>
<p>

#### This is a regression, and the history is specific

Before #26134, `_db_health_readiness_check` called `PrismaDBExceptionHandler.handle_db_exception(e)`, which re-raised only when the flag was off. `health_readiness` caught that and turned it into `HTTPException(503)`. So the flag did gate readiness: with it on you got 200 and `"db": "disconnected"`, with it off you got 503

#26134 (merged 2026-04-20) removed that `handle_db_exception` call to fix a 503 loop. Readiness then returned 200 regardless of the flag

#27003 (merged 2026-05-01) re-added a 503 for the unreachable-DB case, but without any flag check

The net effect is that the flag lost its influence over readiness and the default became "always take the pod out of rotation", which is the opposite of what an operator who sets the flag is asking for

#### Secondary problem: the probe can fail by timeout, not just by status code

`PrismaClient.health_check` is wrapped in `@backoff.on_exception(backoff.expo, Exception, max_tries=3, max_time=10)` (`litellm/proxy/utils.py:4931`) and the underlying `query_raw` has no timeout of its own. A single readiness probe can therefore burn up to ten seconds of retries, or block indefinitely if a connection hangs during a failover, which exceeds the chart's `timeoutSeconds: 5`. Changing only the status code would not be enough, because the probe still fails on time

That work is also duplicated. `_db_health_watchdog_loop` (`litellm/proxy/utils.py:4907`) already runs a bounded `SELECT 1` under `asyncio.wait_for` on an interval and owns reconnection, so the readiness path does not need to drive recovery itself

#### What I expected to happen

With `allow_requests_on_db_unavailable: true`, `/health/readiness` should keep the pod in rotation during a database outage and report the real DB state in the response body rather than through the status code. The DB lookup on the probe path should be bounded so it cannot fail the probe by timeout. The default, with the flag unset or false, should stay exactly as it is today so #27003 is not reverted

A worker that has never successfully loaded its router should still fail readiness even with the flag on, since "router state was never populated" is a different condition from "the DB is momentarily unreachable". That would also address #26237, which asks for stricter readiness in precisely that case

#### Related

#21158 proposed this knob as `LITELLM_READINESS_FAIL_ON_DB_DISCONNECT` and was closed unmerged over a CLA signature and a stale `poetry.lock`, not on the merits. #26237 is the open request pulling in the stricter direction. #14739 and #18415 are the earlier reports that motivated #21158

Happy to open a PR for this if the approach above sounds right

</p>
</details> 

### Steps to Reproduce

1. Start a Postgres you can stop later: `docker run -d --name litellm-db -e POSTGRES_PASSWORD=pw -p 5432:5432 postgres:16`
2. Create `config.yaml`:

```yaml
model_list:
  - model_name: gpt-4.1
    litellm_params:
      model: openai/gpt-4.1
      api_key: os.environ/OPENAI_API_KEY

general_settings:
  allow_requests_on_db_unavailable: true
```

3. Run the proxy: `DATABASE_URL=postgresql://postgres:pw@localhost:5432/postgres litellm --config config.yaml --port 4000`
4. Confirm the baseline: `curl -i http://localhost:4000/health/readiness` returns `200` with `{"status":"healthy","db":"connected"}`
5. Take the database down: `docker stop litellm-db`
6. Call readiness again: `curl -i -w '\ntime_total=%{time_total}\n' http://localhost:4000/health/readiness`. It returns `503` even though the flag is on. The `time_total` also reflects the un-timed retry loop described above
7. Confirm the request layer is still willing to serve: `curl -i -X POST http://localhost:4000/v1/chat/completions -H "Authorization: Bearer sk-1234" -H "Content-Type: application/json" -d '{"model":"gpt-4.1","messages":[{"role":"user","content":"hi"}]}'`. This succeeds, which is exactly what the flag promises
8. On Kubernetes, step 6 failing three times at `periodSeconds: 10` takes every replica out of the Service endpoints, so step 7 becomes unreachable from outside the cluster and a recoverable DB blip turns into a total outage

### Relevant log output

```shell

```

### What part of LiteLLM is this about?

Proxy

### What LiteLLM version are you on ?

v1.95.0

### Twitter / LinkedIn details

_No response_

### PR DESCRIPTION
**Pull request by Zerohertz**
## TLDR

Problem this solves: 

- `allow_requests_on_db_unavailable` had no effect on readiness
- A DB outage marked every replica NotReady in ~30s
- Kubernetes emptied the Service, so the fail-open never ran
- The probe could also blow its timeout on a hung DB

How it solves it:

- Readiness keeps its 200 when the flag is set
- Body still reports the real db status
- The DB lookup is bounded when serving without a DB
- Behaviour with the flag unset is unchanged

## Relevant issues

Fixes #34934

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [X] I have added meaningful tests
- [x] My PR passes all CI/CD checks (e.g., lint, format, unit tests)
- [x] My PR's scope is as isolated as possible; it only solves 1 specific problem
- [x] I have received a Greptile **Confidence Score of at least 4/5** before requesting a maintainer review (Greptile reviews automatically once the PR is opened; only comment `@greptileai` to re-request a review after pushing changes)

## Delays in PR merge?

If you're seeing a delay in your PR being merged, ping the LiteLLM Team on [Slack (#pr-review)](https://join.slack.com/t/litellmossslack/shared_invite/zt-3o7nkuyfr-p_kbNJj8taRfXGgQI1~YyA).

## Screenshots / Proof of Fix

Before is captured at `daf22ec8712e406f502da4268efcb1743d5564ff` (the branch point), after at `771c09f6c04b48f0168abf17d9899e70eed974a0`. Every call below hits a real OpenAI model and costs real money; nothing is mocked

1. Start a Postgres the proxy can lose on demand:

```bash
docker run -d --name litellm-db -e POSTGRES_PASSWORD=pw -p 5432:5432 postgres:16
```

2. Write `proof_config.yaml`:

```yaml
model_list:
  - model_name: gpt-5.6
    litellm_params:
      model: openai/gpt-5.6
      api_key: os.environ/OPENAI_API_KEY

general_settings:
  master_key: sk-1234
  allow_requests_on_db_unavailable: true
```

3. Check out the branch point and start the proxy:

```bash
git checkout daf22ec8712e406f502da4268efcb1743d5564ff
DATABASE_URL=postgresql://postgres:pw@localhost:5432/postgres \
  python litellm/proxy/proxy_cli.py --config proof_config.yaml --detailed_debug \
  --use_v2_migration_resolver 2>&1 | tee before.log
```

4. Confirm the baseline in another shell:

```bash
curl -s -w '\nhttp=%{http_code} time=%{time_total}\n' http://localhost:4000/health/readiness
```

5. Take the database down:

```bash
docker stop litellm-db
```

6. Call readiness again. Before the fix this answers `503` even though the flag is on:

```bash
curl -s -w '\nhttp=%{http_code} time=%{time_total}\n' http://localhost:4000/health/readiness
```

7. Show the request layer is still perfectly willing to serve, which is what makes the 503 a self-inflicted outage:

```bash
curl -s -w '\nhttp=%{http_code}\n' -X POST http://localhost:4000/v1/chat/completions \
  -H 'Authorization: Bearer sk-1234' -H 'Content-Type: application/json' \
  -d '{"model":"gpt-5.6","messages":[{"role":"user","content":"reply with exactly: proof of fix"}]}'
```

8. Simulate the failover case where connections hang instead of being refused, then call readiness once more. Before the fix this outlasts a `timeoutSeconds: 5` probe:

```bash
docker start litellm-db && sleep 5 && docker pause litellm-db
curl -s -w '\nhttp=%{http_code} time=%{time_total}\n' http://localhost:4000/health/readiness
```

9. Stop the proxy, move to this branch, and start it again with the database still paused:

```bash
docker unpause litellm-db && docker stop litellm-db
git checkout litellm_readiness_honor_db_unavailable_flag
DATABASE_URL=postgresql://postgres:pw@localhost:5432/postgres \
  python litellm/proxy/proxy_cli.py --config proof_config.yaml --detailed_debug \
  --use_v2_migration_resolver 2>&1 | tee after.log
```

10. Readiness now answers `200` with the real db status in the body, so the pod stays in the Service:

```bash
curl -s -w '\nhttp=%{http_code} time=%{time_total}\n' http://localhost:4000/health/readiness
```

11. Repeat the hung-connection case and confirm the probe returns inside the bound rather than waiting on the DB:

```bash
docker start litellm-db && sleep 5 && docker pause litellm-db
curl -s -w '\nhttp=%{http_code} time=%{time_total}\n' http://localhost:4000/health/readiness
```

12. Bring the database back and confirm readiness reports `connected` again:

```bash
docker unpause litellm-db
sleep 20 && curl -s -w '\nhttp=%{http_code}\n' http://localhost:4000/health/readiness
```

13. Confirm the default is untouched. Set `allow_requests_on_db_unavailable: false` in `proof_config.yaml`, restart the proxy, stop the database, and check that readiness goes back to `503`:

```bash
docker stop litellm-db
curl -s -w '\nhttp=%{http_code}\n' http://localhost:4000/health/readiness
```

## Type

🐛 Bug Fix

## Changes

`/health/readiness` returned `503` whenever a configured Prisma DB was unreachable and never looked at `general_settings.allow_requests_on_db_unavailable`. Under the probe defaults the Helm chart ships (`readinessProbe.path: /health/readiness`, `periodSeconds: 10`, `failureThreshold: 3`) a database outage marks every replica NotReady inside about thirty seconds, Kubernetes drops all of them from the Service endpoints, and the request-layer fallback that the flag exists to provide never gets a request to run on. A recoverable blip becomes a total outage

The flag used to gate this. `_db_health_readiness_check` called `PrismaDBExceptionHandler.handle_db_exception`, which re-raised only when the flag was off, and `health_readiness` turned that into an `HTTPException(503)`. #26134 dropped the call to fix a 503 loop, then #27003 re-added an unconditional 503, which left the flag with no say over readiness

`_apply_db_readiness_status` now makes that call in one place and both readiness paths share it, so the low-detail probe and the detailed payload reachable through `allow_public_health_readiness_details` cannot disagree. With the flag set an unreachable DB keeps the 200; the `db` field still carries the truth either way

Status code alone is not enough. `PrismaClient.health_check` retries under `backoff` over a `query_raw` that has no timeout of its own, so an unreachable DB can outlast the kubelet's `timeoutSeconds` and fail the probe on time whatever we would have returned. `_readiness_db_status` therefore bounds the lookup with `asyncio.wait_for` when the flag is set. Giving up early costs no recovery, since `_db_health_watchdog_loop` already owns reconnection and runs on its own schedule. The bound deliberately does not apply when the flag is unset, so a slow but reachable DB on the default path behaves exactly as it does today

The budget for that bound is the watchdog's own `PRISMA_HEALTH_WATCHDOG_PROBE_TIMEOUT_SECONDS`, surfaced as a `PrismaClient.db_health_probe_timeout_seconds` property rather than a new setting. Both are answering the same question, how long to wait on a DB liveness check, and an operator who tightens one wants the other tightened too. Note that the 5 second default is longer than some probe configurations allow, so a deployment running `readinessProbe.timeoutSeconds` below that should lower this knob

Four tests cover this. Three fail if the source change is reverted, and the fourth fails if the bound is applied unconditionally, so the "unchanged by default" half of the contract is pinned too. They assert on the returned status and body rather than on elapsed time, so there is nothing timing-sensitive to flake in CI

One thing is deliberately left out of scope. #26237 asks that a worker which never successfully loaded its router fail readiness, and it should keep failing even with this flag set, since "router state was never populated" is a different condition from "the DB is momentarily unreachable". That needs first-successful-load tracking in the startup path, so it belongs in its own PR

### Final Attestation

- [x] The tests check the right things, including the edge cases, and regressions in the respective real-world customer use-cases are not possible after this PR

### CODE REVIEW

File: litellm/proxy/health_endpoints/_health_endpoints.py [hunk 1]
```diff
@@ -1314,6 +1314,79 @@ async def _db_health_readiness_check():
...
+    """
+    return max(
+        0.1,
+        float(
+            os.getenv(
+                "LITELLM_READINESS_DB_PROBE_TIMEOUT_SECONDS",
+                _DEFAULT_READINESS_DB_PROBE_TIMEOUT_SECONDS,
+            )
+        ),
+    )
```
[the remark this answers is not in the document]
Comment (Zerohertz) [reply litellm/proxy/health_endpoints/_health_endpoints.py]: ffc97940108cab371320397df81d4e3ee403ef9d

**Review by Solaris-star — COMMENTED** [review]
Reviewed this in depth — it's the more complete version of the readiness/HA-flag fix (I had attempted a narrower variant that only consulted the flag, so this supersedes that work). A few observations:

**The timeout bound is the part that actually matters, and it's correct**
Consulting `allow_requests_on_db_unavailable` alone isn't sufficient: `PrismaClient.health_check` retries under backoff over a `query_raw` with no timeout of its own, so an unreachable DB can outlast the kubelet's `timeoutSeconds` and fail the probe on *time* regardless of the status code we'd return. Bounding the lookup with `asyncio.wait_for` and reporting `disconnected` on timeout is what keeps the pod genuinely in rotation. Good catch.

**The timeout asymmetry is intentional and right**
The bound is only applied when the fallback flag is set; when it's off, `_readiness_db_status` awaits the check unbounded. That's correct — when fallback is disabled we *want* 503, and a probe that hangs until kubelet times out still yields NotReady, which is the desired outcome. Only the "must return 200 fast" path needs the budget.

**Reusing the watchdog's probe budget is a nice touch**
Exposing `db_health_probe_timeout_seconds` and sharing it between the watchdog and the readiness probe means an operator tightening one tightens both — they answer the same question (how long to wait on a DB liveness check). The docstring spells out exactly this rationale.

**Tests**
The coverage is solid: stays-ready-when-unreachable, the parametrized detailed-payload path (200/503), and the slow-client/probe-timeout test that proves the bound actually fires. The `_slow_prisma_client(delay=0.3, timeout=0.05)` case is a clean way to exercise the timeout without real network flakiness.

One minor, non-blocking thought: on the timeout path `asyncio.wait_for` cancels the in-flight `health_check` mid-backoff. In the test that's a mock so it's moot; in production asyncpg/prisma generally handle cancellation cleanly, but it's the one spot I'd keep an eye on under real DB-outage conditions.
